File: apps/api/src/conversation.ts
import type { Recommender } from "./agent.js";
import {
  isClarify,
  type ClarifyReply,
  type InputItem,
  type Reply,
  type UserContext,
} from "./types.js";
export type Resolve = (said: string[]) => Promise<UserContext>;
export class Conversation {
  constructor(
    private readonly recommender: Recommender,
    private readonly userId: string,
    private readonly resolve: Resolve,
    readonly maxTurns = 3,
    private readonly numItems = 3,
  ) {
    if (maxTurns < 1) throw new Error("max_turns");
  }
  async send(
    said: string[],
    history: InputItem[],
  ): Promise<[Reply, InputItem[]]> {
    const request = {
      user_id: this.userId,
      num_items: this.numItems,
      context: await this.resolve(said),
    };
    const reply = await this.recommender.respond(request, history);
    if (!isClarify(reply)) return [reply, [...history]];
    return [
      reply,
      [
        ...history,
        { role: "user", content: JSON.stringify(request) },
        {
          role: "assistant",
          content: JSON.stringify({
            action: "clarify",
            question: reply.question,
          }),
        },
      ],
    ];
  }
  async converse(
    opening: string,
    ask: (question: string) => Promise<string | null>,
  ): Promise<Reply> {
    const said = [opening];
    let history: InputItem[] = [];
    let last: ClarifyReply | null = null;
    for (let turn = 0; turn < this.maxTurns; turn += 1) {
      const [reply, next] = await this.send(said, history);
      history = next;
      if (!isClarify(reply)) return reply;
      last = reply;
      if (turn === this.maxTurns - 1) break;
      const answer = await ask(reply.question);
      if (answer === null) break;
      said.push(answer);
    }
    if (!last) throw new Error("conversation_failed");
    return last;
  }
}

File: backend/app/conversation.py
from __future__ import annotations

import json
from collections.abc import Awaitable, Callable
from typing import Any

from app.agent import Recommender
from app.models import ClarifyReply, RecommendationRequest, Reply, UserContext

ResolveContext = Callable[[list[str]], Awaitable[UserContext]]


class Conversation:
    def __init__(
        self,
        recommender: Recommender,
        user_id: str,
        resolve_context: ResolveContext,
    ) -> None:
        self.recommender = recommender
        self.user_id = user_id
        self.resolve_context = resolve_context

    async def send(
        self,
        said: list[str],
        history: list[dict[str, Any]],
    ) -> tuple[Reply, list[dict[str, Any]]]:
        request = RecommendationRequest(
            user_id=self.user_id,
            num_items=3,
            context=await self.resolve_context(said),
        )
        reply = await self.recommender.respond(request, history)
        # 推荐是会话终态，不需要再为下一轮保存模型历史。
        if not isinstance(reply, ClarifyReply):
            return reply, list(history)

        # 澄清后用户还会继续回答，因此要把本轮请求与问题都保存成
        # Agents SDK 认识的结构化消息，而不是普通字符串。
        user_history_item = {
            "role": "user",
            "content": [
                {
                    "type": "input_text",
                    "text": request.model_dump_json(),
                }
            ],
        }
        assistant_history_item = {
            "role": "assistant",
            "status": "completed",
            "content": [
                {
                    "type": "output_text",
                    "text": json.dumps(
                        {
                            "action": "clarify",
                            "question": reply.question,
                        },
                        ensure_ascii=False,
                    ),
                }
            ],
        }
        next_history = list(history)
        next_history.append(user_history_item)
        next_history.append(assistant_history_item)
        return reply, next_history
